### ISSUE TITLE
Drawing primitives with screen coordinates

### ISSUE THREAD

**Issue by standard-two-simplex** [gonetz/GLideN64#2324]
I was investigating the wrong texture coordinates issue and wanted to move the vertex position a little, when I realized it was really hard due to the fact that they don't use a viewport transform in the N64. In order to adapt to OpenGL, current master uses the framebuffer size to normalize vertex position and apply its "compulsory" viewport transform.

Inspired on the investigation I did about the clip ratio, I realized that it is possible to draw without using the buffer size. The viewport transform is normally regarded as a function that maps the `[-1,1]x[-1,1]` rectangle into the `[x_offset, x_offset + width]x[y_offset, y_offset+height]` but it is really an affine transformation, i.e., a multiplication and an addition, given by the formula
```
(x,y) -> ((x+1)*width/2+x_offset, (y+1)*height/2 + y_offset)
```

This allows for "original" viewport transforms such as `glViewport(-1,-1,2,2)` which is in fact... the identity function!
```
(x,y) -> ((x+1)*2/2+ (-1), (y+1)*2/2 + (-1)) = (x,y) 
 ``` 
This would mean that screen coordinates would map directly into screen coordinates with such awkward viewport transform. 

There is an obstacle though: primitives are clipped to the `[-w,w]x[-w,w]` boundary, so the natural `w=1` doesn't work with screen coordinates. In order to overcome this, the following can be done,
- Multiply `w` by a big number such as `1024` at the end of the vertex shader, so that screen coordinates aren't clipped. After perspective division the vertex coordinates will be `(x/1024, y/1024)`.
- Use viewport transform given by `glViewport(-1024,-1024,2048,2048)` which is really a multiplication by 1024.
```
(x,y) ->((x+1)*2048/2 +(-1024), (y+1)*2048/2)+(-1024)) = (1024*x,1024*y)
```
All in all, screen coordinates are mapped to screen coordinates.

PROS:
- No dependence on buffer size (see the simplicity of `updateScreenCoorsViewport2()`).
- No divisions except for the perspective division, which may help avoid precision loss.
- N64-like values in vertex shader, which makes it easier to manipulate them.

CONS:
- Lost OpenGLs built in clipping. There should be ways to overcome this though.

I implemented the idea for rectangles and texture rectangles, and they are rendering pretty nicely. On top of that, I added an extra commit trying to correct texture coordinates to illustrate how it is simpler to move vertices, and easier to disable for incompatible features like multisampling (at least much less tangled than the glViewportIndexed() extension). It is still not polished and some features are not supported yet (test without native texrects), but [this branch](https://github.com/standard-two-simplex/GLideN64/tree/use_screen_coords) serves as proof of concept. 

I'm trying to implement a similar idea for triangles, because it is proving much more difficult to hit correct texture coordinates. It is more difficult than for rectangles, because the vertices may or may not go through transformations in the RSP. From what I understand, they are transferred to the RDP in screen coordinates, so the idea would be to apply the viewport transform in the vertex shader when neccessary, then do the same as with rectangles.

@gonetz Can you help me understand the vertex pipeline? I need to know at least...
- What does the vertex attribute `aModify` contain exactly?
- Are vertex coordinates (the ones loaded into the vertex shader) sometimes in screen space and sometimes in device coordinates? If so, how does one tell when is which one?

**Comment by standard-two-simplex** [comment]
It's starting to look good! Native res looks awesome. Higher resolutions are not perfect, but at least look similar to master, I believe. There are issues with bilinearly filtered tiled textures (like in Mario Kart).

TODO
- Native texrects.
- Investigate misalignments with framebuffer textures.
- Review texture coordinate correction in higher resolutions.
- Implement poligon clipping???

 It seems scissoring removes any undesired geometry anyway.

Some screenshots (native):
<table>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92325394-9d911a80-f04a-11ea-8390-14e98ba95cab.png" width = 320>
</td>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92325414-d5985d80-f04a-11ea-85ba-3b1227900954.png" width = 320>
</td>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92325417-d630f400-f04a-11ea-999e-466f009e568e.png" width = 320>
</td>
</tr>
<tr>
<td> BAR - correct HUD </td><td> Star Fox- correct HP bar</td> <td> Ogre Battle - correct dialog boxes </td>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92325418-d6c98a80-f04a-11ea-8d4e-cea6ae2ebcd4.png" width = 320>
</td>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92325419-d6c98a80-f04a-11ea-8230-d953049e7e62.png" width = 320>
</td>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92325620-6a4f8b00-f04c-11ea-92ea-1901a6183dff.png" width = 320>
</td>
</tr>
<tr>
<td> Pilotwings - correct minimap </td><td> Resident Evil - correct options menu</td> <td> SM64- correct health bar </td>
</tr>
</table>

**Comment by standard-two-simplex** [comment]
> Review texture coordinate correction in higher resolutions.

The vertex shift doesn't seem as effective in higher resolutions, because there are more pixels and the first sample is closer to the edge anyway. Instead, I applied a texture coordinate shift and an heuristic for negative deltas to render texrects in a similar fashion to master in higher resolutions. 

My intention was to apply this to every primitive, but some triangles don't seem to benefit from it, so I'm not sure whether is worth it.

Still have to check the behaviour with multisampling enabled.

> Investigate misalignments with framebuffer textures.

I think there are two issues.
- For one, the textures contain garbage at the bottom and sometimes in the right. If you run the games with framebuffer emulation off, it will seem the garbage is unintended.
  - If I subtract 10.0 from texture coordinates, a transparent pixel is clamped and the shadow is moved but shows no garbage.
  - If I add 10.0 to the texture coordinates garbage starts showing. It seems there are many lines of garbage at the bottom (y coordinate) and possibly one line of garbage is begin clamped on the right.
<table>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92520742-62dedc00-f21c-11ea-8687-45a10b3e5fbf.png" width=320>
</td>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92520739-62464580-f21c-11ea-8260-069c1a4a6eee.png" width=320>
</td>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92521357-658e0100-f21d-11ea-8449-6d6f2fd30925.png" width=320>
</td>
<td>
<img src="https://user-images.githubusercontent.com/12202580/92521353-64f56a80-f21d-11ea-9faa-9d50a604c240.png" width=320>
</td>
</tr>
<tr>
<td> Subtracting 10.0 </td> <td> Adding 10.0 </td>
</table>
- Texture coordinates seem to be off by one in the `y` direction. This might be due to a missing line in the top left for one extra in the bottom right, but I can't be sure. It may just be that the texture coordinate has to be clamped to transparent.

Clamping to the appropiate boundary or not loading the garbage part of the texture is probably necessary to emulate framebuffer effects correctly. I believe it is unrelated to vertex positioning, so I'll add a temporary workaround (similar to an existing one) by subtracting 1.0 to texture coordinates. Garbage still appear in JFG when you jump. However, the texture boundary issue has to be examined.

> Native texrects

I'm clueless here. Some rectangles seem to be in place, others completely out of place. Nothing I tried seemed to send me in the right direction.

**Comment by standard-two-simplex** [comment]
I reviewed multisampling code, it should work as master.

There are some issues with negative viewports. This implementation performs the viewport transform in the vertex shader, which has no problem with negative values. Some code was added to overcome OpenGL's limitations of negative viewports, which is confilcting with the fact that I use original `vscale` and `vtrans` values. I would appreciate any help searching the code that was added for it.

**Comment by standard-two-simplex** [comment]
Fixed negative viewports. I had overlooked the fact that clipping bounds were switched with negative bounds, so all geometry was being clipped away.

Some cleanup added.

**Comment by gonetz** [comment]
Hi!

Sorry for the late answer. I'm swamped with my work :(

Your idea sounds very interesting and the results look promising.

> What does the vertex attribute aModify contain exactly?

Flags, which parts of vertex are already pre-calculated.
gSP.h:
```
#define MODIFY_XY	0x000000FF
#define MODIFY_Z	0x0000FF00
#define MODIFY_ST	0x00FF0000
#define MODIFY_RGBA	0xFF000000
#define MODIFY_ALL	0xFFFFFFFF
```

gSPModifyVertex sets these flags. Also drawScreenSpaceTriangle sets MODIFY_ALL.

> Are vertex coordinates (the ones loaded into the vertex shader) sometimes in screen space and sometimes in device coordinates?

Yes.

> If so, how does one tell when is which one?

Set corresponding flag in vtx.modify 


I'll check Native texrects after I sort out my urgent work tasks.

Good luck!

**Comment by standard-two-simplex** [comment]
The task is nearly completed, but I have some small precision issues in HLE.

<table>
<tr>
<td> <img src="https://user-images.githubusercontent.com/12202580/93375771-f2a60b00-f858-11ea-83b5-84448ab6345f.png" width="320px" /></td>
<td><img src="https://user-images.githubusercontent.com/12202580/93375773-f33ea180-f858-11ea-8efb-bfd5ca13063c.png" width="320px" /></td>
</tr>
<tr>
<td> <img src="https://user-images.githubusercontent.com/12202580/93375775-f33ea180-f858-11ea-9159-6e452f01c696.png" width="320px" /></td>
<td><img src="https://user-images.githubusercontent.com/12202580/93375776-f3d73800-f858-11ea-913d-7c3f9946cc67.png" width="320px" /></td>
</tr>
<tr>
<td> <img src="https://user-images.githubusercontent.com/12202580/93375777-f3d73800-f858-11ea-9ec8-c5247368c8f0.png" width="320px" /></td>
<td><img src="https://user-images.githubusercontent.com/12202580/93375778-f46fce80-f858-11ea-9dcb-7256cc3b5824.png" width="320px" /></td>
</tr>
<tr>
<td> HLE </td><td>LLE</td>
</tr>
</table>

I checked Ogre Battle's case. This are the vertex position coordinates I got in HLE of one of the vertices involved.

Before matrix transformations
```
x: 112.000000, y: 40.000000, z:0.000000 
```
After multiplication by combined matrix.
```
x: -0.6989746093750000, y: 0.3332519531250000, z:-0.6364440917968750, w:1.0000000000000000 
```
After viewport transform
```
sx: 48.1640625000000000, sy:80.0097656250000000
```

If those were 48.0 and 80.0 the scene would render correctly.  I still need to find a way to get this values in LLE. Since the scene renders correctly, the RSP plugin must get different values which are then passed via a screen coordinate draw call. However, I can't image what can possibly go wrong with the matrix multiplication.

**Comment by olivieryuyu** [comment]
matrix multiplication in RSP are done in 16 bits and there could be some small imprecisions.

If you give me a PJ64 save state and an accurate log of HLE commands (including the exact vtx and tri command you want to check), i can have a look to the results of the vertex processing.

**Comment by standard-two-simplex** [comment]
GLideN64 reports command `gSPVertex(2150084864,4,0)` and `gSPVertex(2150084868,4,0)`
The four vertices at that address should have the following coordinates.
```
gsP(2150084864,4,0) 
x: -112, y: -52, z:0 
x: -112, y: 40, z:0 
x: 112, y: 40, z:0 
x: 112, y: -52, z:0 
```
Savestate: [Ogre Battle 64 - Person of Lordly Caliber (U) (V1.0) [!].zip](https://github.com/gonetz/GLideN64/files/5234888/Ogre.Battle.64.-.Person.of.Lordly.Caliber.U.V1.0.zip)

I don't know the triangle call. If you need anything more, tell me and I'll try to figure it out.

**Comment by olivieryuyu** [comment]
this is not a log of HLE commands, I would know with that which commands are generating the bug.

@gonetz 

Would u be able to help here?

**Comment by gonetz** [comment]
The background in this DATA window is rendered by this fragment of display list:
```
0x0028F608: CMD=0xFD W0=0xFD500000 W1=0x8024DB20
gDPSetTextureImage( G_IM_FMT_CI, G_IM_SIZ_16b, 1, 0x0024DB20 );
0x0028F610: CMD=0xF5 W0=0xF5500000 W1=0x07018060
gDPSetTile( G_IM_FMT_CI, G_IM_SIZ_16b, 0, 0, 7, 0, G_TX_NOMIRROR, G_TX_NOMIRROR, 6, 6, 0, 0 );
0x0028F618: CMD=0xE6 W0=0xE6000000 W1=0x00000000
Ignored: gDPLoadSync();
0x0028F620: CMD=0xF3 W0=0xF3000000 W1=0x073FF200
gDPSetTileSize( 7, 0.00, 0.00, 255.75, 128.00 );
gDPLoadBlock( 7, 0, 0, 1023, 512 );
0x0028F628: CMD=0xE7 W0=0xE7000000 W1=0x00000000
Ignored: gDPPipeSync();
0x0028F630: CMD=0xF5 W0=0xF5400800 W1=0x00018060
gDPSetTile( G_IM_FMT_CI, G_IM_SIZ_4b, 4, 0, 0, 0, G_TX_NOMIRROR, G_TX_NOMIRROR, 6, 6, 0, 0 );
0x0028F638: CMD=0xF2 W0=0xF2000000 W1=0x000FC0FC
gDPSetTileSize( 0, 0.00, 0.00, 63.00, 63.00 );
0x0028F640: CMD=0x01 W0=0x01004008 W1=0x8028B0E0
gSPVertex n = 4, v0 = 0, from 8028b0e0
0x0028F648: CMD=0x06 W0=0x06000204 W1=0x00000406
gSP2Triangle (0, 1, 2)-(0, 2, 3)
```
The whole log attached
[gliden64.debug.log.zip](https://github.com/gonetz/GLideN64/files/5237168/gliden64.debug.log.zip)

**Comment by gonetz** [comment]
Vertices coordinates:
![OB64-vtx-coords](https://user-images.githubusercontent.com/8983092/93437236-f340af00-f8f5-11ea-90f8-afcc810cb6d7.png)

**Comment by olivieryuyu** [comment]
may i get a PJ64 savestate?

Mupen wont't work for me for my purpose

**Comment by standard-two-simplex** [comment]
Here: [Ogre Battle 64 - Person of Lordly Caliber (U) (V1.0).pj.zip](https://github.com/gonetz/GLideN64/files/5241628/Ogre.Battle.64.-.Person.of.Lordly.Caliber.U.V1.0.pj.zip)

**Comment by olivieryuyu** [comment]
Here my what I can get about the vtx processing.

[Analysis.xlsx](https://github.com/gonetz/GLideN64/files/5247591/Analysis.xlsx)

You have the 4x4 MP matrix in 16 bits

the vertices in 16 bits

after multiplication in 16 bits you get vertex structures which is then used by the tri commands to generate LLE triangles.

The vtx structures after MP matrix matches the LLE results.

Xscrn and Yscrn, which you want to get I believe, is from my understanding S11.2 format.

I have the same results than you for LLE.

**Comment by gonetz** [comment]
It is not very hard to replace floating point matrix-matrix and vector-matrix multiplication by fixed point one.

**Comment by olivieryuyu** [comment]
There also the multiplication of viewport structure. This is how you actually get the S11.2 format.

@standard-two-simplex 

Hope it helps to understand the imprecision you have in HLE.

**Comment by standard-two-simplex** [comment]
Thanks, I'll check HLE values and compare to them.

**Comment by standard-two-simplex** [comment]
The combined matrix in HLE is
```
m[0][0]: -0.0062408447265625, m[0][1]: 0.0000000000000000, m[0][2]: 0.0000000000000000, m[0][3]: 0.0000000000000000 
m[1][0]: 0.0000000000000000, m[1][1]: 0.0083312988281250, m[1][2]: 0.0000000000000000, m[1][3]: 0.0000000000000000 
m[2][0]: 0.0000000000000000, m[2][1]: 0.0000000000000000, m[2][2]: 0.0181732177734375, m[2][3]: 0.0000000000000000 
m[3][0]: 0.0000000000000000, m[3][1]: 0.0000000000000000, m[3][2]: -0.6364440917968750, m[3][3]: 1.0000000000000000 
```
In fact, it is the same as the projection matrix and the model matrix is the identity matrix. Observe that the values on the diagonal are close to `-1.0/160.0` and `1.0/120.0` where the denominators are the scale values of the viewport transform.

@olivieryuyu How do I interpret this matrix? I'm guessin the higher 16 bits hold the integral part and the lower 16bits the fractional part. But how are they ordered?
```
4x4 MP matrix
--
0xFFFF0000
0x00000000
0x00000000
0x00000000
0x00000000
0x00000000
0x00000000
0xFFFF0001
0xFE670000
0x00000000
0x00000222
0x00000000
0x00000000
0x04A70000
0x00000000
0x5D120000
```

**Comment by olivieryuyu** [comment]
I believe this is explained here:

http://n64devkit.square7.ch/n64man/gsp/gSPMatrix.htm

**Comment by standard-two-simplex** [comment]
> Xscr and Yscr, which you want to get I believe, is from my understanding in S11.2 format.

Yes, this are the ones I want to get accurately. From gonetz's pictures, it seems they should be 48.0 and 80.0 for this given vertex.

> The vtx structures after MP matrix matches the LLE results.

True, so the problem lies only after the viewport transform.

> Xscr and Yscr, which you want to get I believe, is from my understanding in S11.2 format.

Yes. I had tried rounding to the closest 1/4 of an integer, but it seems that I should simply drop the data after the third bit, simulating a simple bit shift. It is backed up by the fact that the other vertex with coordinate `171.983` in HLE is at position `171.75` instead of the much closer `172.0`.

**Comment by standard-two-simplex** [comment]
So rounding down to a 1/4 integer seems to solve the three issues above.

I simply applied a correction to position data.
```
gl_Position.xy = floor(gl_Position.xy * vec2(4.0)) * vec2(0.25);
```

However, it might be nicer to do all computations in integers and simply apply the correct subpixel mask to the result.

**Comment by gonetz** [comment]
> However, it might be nicer to do all computations in integers

All - matrix, vertex and viewport transformations?

**Comment by standard-two-simplex** [comment]
> All - matrix, vertex and viewport transformations?

Yes that's what I meant. But I'm not completely sure about the viewport transformation.

I don't know how the N64 handles perspective division. Does the RSP perform division by `w` and send `1/w` over to the RDP for perspective corrected interpolation of attributes? Or does it send all four `x`, `y`, `z` and `w` and then the RDP performs the division and calculation of `1/w`?

Currently, I am handling the viewport transform in this way in the vertex shader.
```
gl_Position.xy = gl_Position.xy * vec2(1.0, -1.0) * uVScale + uVTrans * vec2(gl_Position.w);
```
Ignore the `vec2(1.0,-1.0)` for now, the `y` coordinate seems to be inverted. The last `w` is necessary so that after OpenGL performs perspective division we get
```
(gl_Position.xy/gl_Position.w) * uVScale + uVTrans
```
The same could be done in `processVertex()` in fixed point arithmetic and the result stored in s11.2 format, but I wonder if this is the way the RSP behaves.

**Comment by olivieryuyu** [comment]
Vertex processing is done in this way, at least for Fast3D family ucode:

1)    Multiply vertices with ModelViewProjection matrix
2)    Perform lightings and texture coordinates transformation tasks.
3)    Determine the outcodes of the vertex for clipping (https://en.wikipedia.org/wiki/Cohen-Sutherland_algorithm)
4)    Normalized device coordinates (divide everything by W)
5)    Transform clip coordinates to screen coordinates, including perspective correction (viewport transformation)
6)    Calculate fog where necessary
7)    Store results in the DMEM vertex buffer

All calculations are done in 16 bits

**Comment by standard-two-simplex** [comment]
> 5. Transform clip coordinates to screen coordinates, including perspective correction (viewport transfromation)

What do you mean "including perspective correction" ?

**Comment by standard-two-simplex** [comment]
I was concerned about perspective correct interpolation if we divided by `w`, so I reviewed the theory. I think it is possible to do the following.
1. Multiply vertices with ModelViewProjection matrix. (operations in integers)
3. Determine the clip distances `w - x`, `x - (-w)`, `w-y`. `y- (-w)` and load them as vertex attributes. OpenGL can use them in a similar fashion to the algorithm given above. Alternatively, this step can be skipped and the distances calculated in vertex shader.
4. Normalized device coordinates (divide everything by W) (operations in integers). (`w` is 1 at this step, so perspective correction would be lost with these coordinates)
5. Perform the view transform (operations in integers). Apply the correct bit shift to u11.2 format, dropping some of the fractional part.
6. Cast the coordinates to float to load them into the vertex shader. If perspective correction is enabled, multiply the coordinates by `w` (due to the precision loss in step 5, the result is different from simply skipping step 4 and 6) (operations in floats).

OpenGL will perform division by `w` again, and so the vertices will end up with the expected screen coordinates. Besides, if `w` has the original value, perspective correction will be applied. If `w` is 1, perspective correction won't be applied.

I need to cleanup the current WIP history, it's a real mess. After that I'll try to implement the above idea.

**Comment by olivieryuyu** [comment]
may be I express myself improperly on perspective correction: 

Please see 11.3.3 Perspective Normalization in the N64 programming manual for explanation

**Comment by standard-two-simplex** [comment]
> All - matrix, vertex and viewport transformations?

I checked the code and realize that this is a lot of work. Matrix calculations are everywhere in vertex code and changing the data to integers means changing every vertex operation.

Instead I followed the guideline I proposed above (except for clipping) but performed operations on floats. The `vtx.x = floorf(vtx.x* 4.0f) *0.25f;` looks ugly, but at least I managed to extract the vertex processing outside the vertex shader and all vertices (loaded via gSPVertex or gSPModifyVertex) go through the same vertex shader (no branches).

I need to review clipping. Do you guys know if/how the vertices compute new outcodes if new x,y data is loaded via gSPModifyVertex() ?

**Comment by olivieryuyu** [comment]
what do you mean by new outcodes?

gSPModifyVertex modifies directly a portion of the vertex previously calculated and stored in DMEM with GSPVertex.

So if you change Xscrn or Yscrn, you have directly the right info for your purpose.

**Comment by standard-two-simplex** [comment]
> what do you mean by new outcodes?

In order to decide whether a vertex has to be clipped, it has to be compared to certain boundaries. Usually this happens at clipping coordinate stage, where x,y,z are compared to -w and w.

If I modify its Xscrn or Yscrn directly, I am bypassing the comparison step. I don't know what happens to such vertices. Are they assumed to be inside the clipping space?

**Comment by olivieryuyu** [comment]
never thought about this

the clipping is done on the base of the previous clipping data stored in DMEM for the vertex modified.

**Comment by standard-two-simplex** [comment]
> the clipping is done on the base of the previous clipping data stored in DMEM for the vertex modified.

Ok, thanks. I guess this should be simple to implement then.

I was checking an issue I was having with fog and came across this piece of code.
```
"    if (aPosition.z < -aPosition.w && aModify[1] == 0.0)  \n"
"      fp = -uFogScale.s + uFogScale.t;                        \n"
"    else                                                      \n"
"      fp = (aPosition.z/aPosition.w)*uFogScale.s + uFogScale.t;             \n"
```

I can follow the else part. But what does the if part do?

**Comment by olivieryuyu** [comment]
that would be sergey to help out here

@gonetz

**Comment by gonetz** [comment]
> Are they assumed to be inside the clipping space?

Yes. Our code sets the corresponding flag in vtx.modify and then this modified coordinate(s) is multiplied by w in vertex shader.

**Comment by gonetz** [comment]
> I can follow the else part. But what does the if part do?

Clamps aPosition.z/aPosition.w to -1.
It is from original glN64 code.  I tried to remove that part, but it causes regressions.

**Comment by standard-two-simplex** [comment]
I eventually computed clip distances in `processVertex()` and I wanted to load them as vertex attributes so that OpenGL can perform clipping via its built in clip distances.

However I am having some trouble loading the data. I mimicked other vertex attribute loading without understanding well what's going on. Could I get some help with commit 212b435176cc54749a4a792eda93023414721e50 ?

The shader links correctly with the variable `IN hightp vec4 aClipDistances` used, but clipping doesn't happen correctly, so the data is not being uploaded probably.

![mariokart64-000](https://user-images.githubusercontent.com/12202580/95616121-9a35e800-0a69-11eb-9d5d-c234d1204a2d.png)
Clouds not being clipped.

**Comment by standard-two-simplex** [comment]
I think I managed to load the data correctly. The PR is almost ready, the output image is what I was expecting in most cases. Only Native Rects and FB emulation off are unsupported. I will try to adapt them again, but I will probably need help.

It would be great if the feature were teste before it is merged into master. If you (the reader) use this plugin for casual play, I would appreciate if you tested the artifacts in the [pull request](https://github.com/gonetz/GLideN64/pull/2334) and report any regression.
- Native resolution (UseNativeResolutionFactor=1) looks really accurate. I believe it should be possible to achieve perfect alignment. Therefore, any bug present in native res is considered critical.
- Non-native resolution aren't perfect. There are games that contain bugs in their code and render correctly only because the resolution they are rendering in. Therefore, bugs present only in non-native might not be solvable without hacks.
- Alignment issues which are not regressions might not be related, so better to open their own issue and simply mention them here so I can check if they are related.

Thanks for any help with the testing.

**Comment by standard-two-simplex** [comment]
I found a few issues, and diagnosed the problems.

- Vigilante 8 has some vertices with `w=0`. They are meant to be clipped before division by `w`, so that w of the clipped vertices will be that of the near plane. @olivieryuyu 

> Normalized device coordinates (divide everything by W)

Are you sure that is happening in the vertex processing? It seems to me that a primitive has to be generated and vertices clipped before division happens. Is it possible that perspective normalization is happening instead? (i.e., multiplication of x,y,z and w by a constant factor so that w is close to 1 for a point halfway between the near and far planes).

- Star Wars games seem to flip the vertex y coordinate after `processVertex()` is called. Current master code goes,
```
template <u32 VNUM>
u32 gSPLoadSWVertexData(const SWVertex *orgVtx, SPVertex * spVtx, u32 vi, u32 n)
    {
	const u32 end = n - (n%VNUM);
	for (; vi < end; vi += VNUM) {
		for(u32 j = 0; j < VNUM; ++j) {
			SPVertex & vtx = spVtx[vi+j];
			vtx.x = orgVtx->x;
			vtx.y = orgVtx->y;
			vtx.z = orgVtx->z;
			++orgVtx;
		}
		gSPProcessVertex<VNUM>(vi, spVtx);
		for (u32 j = 0; j < VNUM; ++j) {
			SPVertex & vtx = spVtx[vi+j];
			vtx.y = -vtx.y;
		}
	}
	return vi;
}
```

I want to perform the viewport transform inside of process vertex, which conflicts with the y inversion that needs to happen earlier. Is there a way to determine inside `gSPProcessVertex()` if it needs y inversion? Currently I rely on an ugly workaround.

- I am performing clipping in HLE by loading clip distances. However, I need to disable them for LLE because the RSP already processed the vertex. @gonetz Do you know how I can disable this piece of code? My current implementation isn't working. 
```
	if (!RSP.LLE)
			m_part +=
			"  gl_ClipDistance[1] = aClipDistances[0];						\n"
			"  gl_ClipDistance[2] = aClipDistances[1];						\n"
			"  gl_ClipDistance[3] = aClipDistances[2];						\n"
			"  gl_ClipDistance[4] = aClipDistances[3];						\n"
			;
```

**Comment by olivieryuyu** [comment]
In HLE:

From my understanding W is calculated with vertex processing. You can check out the source code of the microcodes if you want to get confirmation.

Perspective normalization is also done.

Clipping is working in the way:

Each vertex has some flag to define in which part of the screen or outside the screen the vertex is.

If all vertex are outside the screen, the triangle is rejected

If at least one of them is in the screen, then clipping may occur.

If all vertex is within the clipratio box, the triangle is fully drawn without any recalculation.

If one of the vertex is outside the clipratio box, then clipping occurs, meaning that vertices are somehow recalculated to be at the edge of the clipratio box, with potentiall split of the triangles into few triangles. 

With those new vertices, your reprocess the vertex(es) calculations including the calculcation of W and then draw immediately those triangles (to the RDP).

This is why clipping is a heavy process on RSP and RDP.

For the Y inversion, this is a HLE matter. The "natural" coordinates of the N64 are with this Y inverted actually. Most of microcode do a small calculation to revert those coordinates. It thinks it has to do with the coordinates system of the N64.

**Comment by standard-two-simplex** [comment]
> For the Y inversion, this is a HLE matter. The "natural" coordinates of the N64 are with this Y inverted actually. Most of microcode do a small calculation to revert those coordinates to match with OpenGL. It thinks it has to do with the coordinates system of the N64.

So I actually managed to perform the viewport transform correctly by checking `GBI.isNegativeY()` and code is much cleaner now. All games I tested are working, except SM64. It seems that `GBI.isNegativeY()` returns false, but the geometry is inverted. Any idea why it is not initialized to true?

**Comment by standard-two-simplex** [comment]
> So I actually managed to perform the viewport transform correctly by checking GBI.isNegativeY() and code is much cleaner now. All games I tested are working, except SM64. It seems that GBI.isNegativeY() returns false, but the geometry is inverted. Any idea why it is not initialized to true?

Actually, there are more games that were also failing. Apparently games use different methods to do the inversion (different combined matrix, negative viewport transform...). Since I changed the way the viewport transform is one, which supports negative scales, I updated the isNegativeY properties on GBI.cpp.

**Comment by standard-two-simplex** [comment]
I updated on the heuristic for non-native rects: if a texture coordinate exceeds the min/max value of the values reached in native resolution, it is clamped to it. As a result, a hard edge is shown where garbage was before.

Non-tiled rectangles look good in non-native now too. Tiled textures are not perfect, but are certainly improved.

![mariokart64-001](https://user-images.githubusercontent.com/12202580/99913997-0d827900-2cfb-11eb-800b-1f845b46ad82.png)
![mariokart64-002](https://user-images.githubusercontent.com/12202580/99913996-0c514c00-2cfb-11eb-9d4d-84d5af5ee928.png)

**Comment by gonetz** [comment]
> Tiled textures are not perfect, but are certainly improved.

True.

**Comment by standard-two-simplex** [comment]
I have taken advantage of the software clipping to overcome the two difficulties I had before. On the one hand, I had dubious clipping code in place, because OpenGL clipping was being hacked (in a similar way as the slingshot hack) to prevent the screen coordinates from being clipped in unintended boundaries. On the other, I had issues with performing the viewport transform before clipping for vertices with w==0. The first problem is now simply gone and the second is overcome by performing the viewport transform in the vertex shader, when the software clipper has already done its work.

As a result, I have a much cleaner implementation. In fact, it is the first time that I am happy with the code. Currently the 6 commits do the following.
- Use screen coordinates for vertex positions: triangles processed by rsp have the viewport transform is performed in the vertex shader and rectangle positions are  simply passed to the vertex shader. The vertex shader divides `x` and `y` by 1024 and OpenGL viewport transform multiplies it back before sending to the rasterizer.
- Workaround for framebuffer textures. It was added for CBFD shadow borders. I changed the hack to adapt it to the new texture coordinate handling. Obviously, it is still not the correct solution. 
- Apply a vertex shift in native and texture coordinate shift otherwise. In native resolutions it is possible to shift the vertices to approximate N64 raster rules quite correctly. In higher resolutions, there are more samples than intended and this method often becomes more of a source of issued than a solution. Instead, I decided to shift texture coordinates and added a heuristic to render different texture rectangles (like the initial logo in Rocket Robot on Wheels) as in master. 
- Use texture coordinate bounds for texrects in non-native resolutions. Minimum and maximum texture coorinates achieved in native are computed and used as bound for non native resolutions. It improves the visual artifacts due to magnification in most cases, but there are a few cases where it looks worse. In view of this, it could be made optional.
- Invert the viewport based on GBI.isNegative() and reasses ucodes with negative viewport. Negative viewport scales are no longer an issue because the multiplication is done in the vertex shader. Therefore, I simplified the viewport inversion and reassesed wich ucodes need inversion.

I believe everything is working fine, except for NativeRects option. I  have not managed to figure out how that code should be adapted. I need some help with it. Some cleanup is also needed: for example, I didn't replace old useScreenCoordsViewport() and instead created useScreenCoordsViewport2() because I wasn't able to remove the reference in TexrectDrawer.cpp. Other than this, I believe the branch is ready for merging, if the changes are wanted.

**Comment by fzurita** [comment]
Any chance this could fix this issue?

https://github.com/gonetz/GLideN64/issues/1952

**Comment by standard-two-simplex** [comment]
> #1952

There are so many different issues reported in that thread. I believe some are fixed in master and I know some others are not fixed in this branch.

**Comment by fzurita** [comment]
Yeah, I know that starcraft has issues in master, at least once you get in game. The sprites have some junk in them.

**Comment by gonetz** [comment]
@standard-two-simplex Sounds cool! I'll try to do the code review asap.

**Comment by standard-two-simplex** [comment]
I reworded limits to bounds to better reflect what the variable does. I also added a config option to enable texture coordinate bounds optionally. It's working in mupen64plus and ready to be plugged for zilmar spec.

**Comment by gonetz** [comment]
Request for everyone: please test the latest build with that feature for regressions:
https://ci.appveyor.com/project/gonetz/gliden64/builds/38285888/artifacts

I'm going to add it to master soon.

**Comment by standard-two-simplex** [comment]
![mariokart64-000](https://user-images.githubusercontent.com/12202580/111661100-04b83480-880f-11eb-8cfe-321372b24206.png)

Mario Kart seems to have broken with changes to clipping in this branch. :´(

**Comment by olivieryuyu** [comment]
eikou st andrews

![GLideN64_����ɾ�ı���ح��_000](https://user-images.githubusercontent.com/8987490/111808341-36470380-88d4-11eb-84c3-fa29801df7fa.png)

**Comment by olivieryuyu** [comment]
![GLideN64_rogue_squadron_002](https://user-images.githubusercontent.com/8987490/111808757-98076d80-88d4-11eb-81a2-4a1ec6940188.png)

**Comment by gonetz** [comment]
SW-RS issue introduced with commit 516978008adf "Invert the y coordinate based on GBI.isNegativeY()"

**Comment by gonetz** [comment]
Clipping is fixed: a74e4709

**Comment by standard-two-simplex** [comment]
Yes, I have been checking the issue. Before some refactors, I was performing the viewport transform in gSP.processVertex() where the changes were working. After the refactors the viewport transform is performed in vertex shader (i.e., after clipping) and this changes stopped working.

In order to restore SW to work it is necessary to restore
```
for (u32 j = 0; j < VNUM; ++j) {
			SPVertex & vtx = spVtx[vi+j];
			vtx.y = -vtx.y;
}
```
and change the negativeY property of the microcode to true.

In order to restore Eikou no St Andrew, it is necessary to restore
```
if (gSP.viewport.vscale[1] < 0.0f && !GBI.isNegativeY())
		gSP.viewport.vscale[1] = -gSP.viewport.vscale[1];
```
but things get a little complicated becaue the negativeY property of the microcode must be set to true. This invalidates the second condition, and writing GBI.isNegativeY() instead breaks other games with negative vscale[1]. It seems that this game need inverting some vertices but not inverting some other vertices. 

The simplest solution is to drop commits 516978008adf460896eb4cecf84137840da80411 and 448fdc5e5931b6116e17253fa854c1b511407a9a, after which everything works as it was doing before.

**Comment by gonetz** [comment]
Ok, I reverted these two commits.

**Comment by standard-two-simplex** [comment]
Ok, I pushed a cleanup commit to use_screen_coords_2 branch, you may squash it to the first commit before merging.

**Comment by gonetz** [comment]
I think it is better you do the new clean branch, drop the draft PR #2334 and create a new one. And please increase shader cache version in some commit.

**Comment by standard-two-simplex** [comment]
Alright, I pushed the new code to branch use_screen_coordinates. I will double check every commit is working as intended later today/tomorrow.

The final code should be the same as in the test build provided above, so just test that one. I bump it here.
https://ci.appveyor.com/project/gonetz/gliden64/builds/38321808/artifacts

**Comment by gonetz** [comment]
I found one bad issue with "enableTexCoordBounds" option:
![texcoordsbounds](https://user-images.githubusercontent.com/8983092/111896066-898e9280-8a49-11eb-94de-45f7bd08bc65.png)
Is it possible to fix?
mupen64plus savestate:
[Paper Mario (U) [!].zip](https://github.com/gonetz/GLideN64/files/6176947/Paper.Mario.U.zip)
Just move right a bit: normally Mario should disappear but with tex coords bounds enabled it does not work.

**Comment by standard-two-simplex** [comment]
> Is it possible to fix?

Yes, it seems the game uses a big darkness texture and changes the s parameter of the texrect command as you move Mario around. The texture has a shift of 15, so it should be accounted for, which I didn't in my implementation.

**Comment by standard-two-simplex** [comment]
Actually it is a bit complicated. I'll explain the details later. Can I get a pj64 savestate to test with angryluons plugin?

**Comment by gonetz** [comment]
> Can I get a pj64 savestate to test with angryluons plugin?

Here: 
[PaperMarioInTheDark.pj.zip](https://github.com/gonetz/GLideN64/files/6181779/PaperMarioInTheDark.pj.zip)

**Comment by standard-two-simplex** [comment]
So this is the situation. The texture used is a 64x64 texture, that mirrors once in each direction and clamps afterwards. So texture coordinates between [0,127] form the transparent circle, and anything outside clamps to a opaque black pixel.

The game changes the s value of the texture rectangle call to move the transparent circle along the screen, based on Mario's position. Most of the time this is rendered correctly, but there is an exception when Mario goes to the edge of the screen and `s` becomes too negative. See this texrect call in such situation (arguments in floating point format). 

```
gDPTextureRectangle(12.000000, 20.000000, 308.000000, 220.000000, 0, -540.000000, -266.000000, 2.000000, 2.000000, 0)
```

As you can see, the `s` coordinate on the left is `-540`. The tile has a shift of 15, so it becomes -1080. But due to fixed point arithmetic, this number overflows (s10.5 means coordinate range between [-1024.00, 1024]) and becomes +56 in current master. Then, the `s` coordinate on the right is computed by adding the rectangle width times dsdx, so 56 + (308-12)*2*2 = 1240.

This happens in GraphicsDrawer::drawTexturedRect()
```
			if (gSP.textureTile[t]->shifts > 10) {
				const u32 shifts = 16 - gSP.textureTile[t]->shifts;
				S = static_cast<s16>(S << shifts);
				shiftScaleS = static_cast<f32>(1 << shifts);
			} else if (gSP.textureTile[t]->shifts > 0) {
				const u32 shifts = gSP.textureTile[t]->shifts;
				S = static_cast<s16>(S >> shifts);
				shiftScaleS /= static_cast<f32>(1 << shifts);
			}
			const f32 uls = _FIXED2FLOAT(S, 5);
			const f32 lrs = uls + offsetX * shiftScaleS;
``` 

As a result the interpolated texture coordinates range between [56, 1240] and most of them are clamped to a dark texture. This makes Mario suddenly completely disappear. What I don't understand is why there isn't a transparent circle on the left of the screen. In angrylions plugin, Mario is visible also in the edges of the screen.

My current computation of texture coordinate bounds conflicts with this, because it isn't considering the integer overflow. Thus, a lot of those big coordinates are clamped to somewhere below (-540+(308-12)*2)*2 = 104, making a fully visible strip.

Removing `S = static_cast<s16>(S << shifts);` and changing the computation of uls to `const f32 uls = (_FIXED2FLOAT(S, 5))*shiftScaleS;` solves this issue in master, and allows the texcoord bounds to work normally. This was introduced in 737338ca2058, so perhaps you remember why it was necessary.

However, From my understanding, the correct behaviour would be to assign unshifted texture coordinates to the vertices and have the rasterizer interpolate them unshifted, and shift them on a per fragment basis in the texture engine. There is a chance that, the N64 RDP uses more than 16 bits at this stage and coordinates don't overflow. But if they do, only the leftmost coordinates will, as the other pixels will have texture coordinates greater than -512.

This is however a bit of a task, so I'm not sure how to proceed. I can probably make it work, but I would need time and it could give me many headaches. Definitely not something to do in the short term.

**Comment by gonetz** [comment]
Thanks for the explanation!

> What I don't understand is why there isn't a transparent circle on the left of the screen. In angrylions plugin, Mario is visible also in the edges of the screen.

It is interesting. Have you checked, how al plugin calculates texture coordinates for that texture on the edge of the screen?

> My current computation of texture coordinate bounds conflicts with this, because it isn't considering the integer overflow.

Why? Is there a problem with that?

> This was introduced in 737338c, so perhaps you remember why it was necessary.

Yes, it was done exactly for integer overflow. In MLB, textures of spectators were corrupted because lack of integer overflow emulation.
Texture coordinates became too large after the shift. Clamp to s16 after the shift fixes the issue.

> Removing S = static_cast<s16>(S << shifts); and changing the computation of uls to const f32 uls = (_FIXED2FLOAT(S, 5))*shiftScaleS; solves this issue in master, and allows the texcoord bounds to work normally. 

Most likely it will bring MLB issue back.

> However, From my understanding, the correct behaviour would be to assign unshifted texture coordinates to the vertices and have the rasterizer interpolate them unshifted, and shift them on a per fragment basis in the texture engine.

Move most of texturing work to fragments, including fetching texels from TMEM, sounds promising. But it is a large and hard work, definitely for the next release cycle.

How to proceed. If it is impossible to adjust computation of texture coordinate bounds such a way that the integer overflow is taken into account, it is pity but ok. This option will be off by default and enabled per-game in custom settings.

**Comment by standard-two-simplex** [comment]
> It is interesting. Have you checked, how al plugin calculates texture coordinates for that texture on the edge of the screen?

No, I only checked the behaviour in Angrylions plugin but didn't dump actual data. I only dumped some data using GLideN64. Actually, I found my mistake, the texture coordinates don't wrap to +56 but to -1080+2048  = 1968. So everything clamps to the black texel on the right of the texture when Mario goes to the right of the screen.

> Why? Is there a problem with that?

No, only that I had written it differently. I updated the bounds computation to consider the shift as in GraphicsDrawer::drawTexturedRect().

This adjusts to current way of shifting texture coordinates, so Mario now completely dissapears with the seeting on or off.

**Comment by standard-two-simplex** [comment]
> From my understanding, the correct behaviour would be to assign unshifted texture coordinates to the vertices and have the rasterizer interpolate them unshifted, and shift them on a per fragment basis in the texture engine. 

I implemented this in https://github.com/standard-two-simplex/GLideN64/tree/per_pixel_shift_offset. I wrote it on top of this branch.

To be honest, I was expecting more regressions, but I only had issues with textures with hdRatio different to one. I think I managed to solve them. It fixes the darkness effect in paper Mario and MLB seems to still be correct.

@gonetz The texture shift might be a bit unrelated to this issue, so if you prefer to have it tested later, there's no problem for me. Or else it can be combined with the current branch. Your call.

**Comment by GhostlyDark** [comment]
Found a regression with the latest build from AppVeyor. Happens with all Minimaps, so you can simply load a save file. The HD textures are accurate, so those aren't an issue. The latest master build looks correct.

HD Minimaps in Majora's Mask:
![image](https://user-images.githubusercontent.com/31742919/113518597-9b168500-9587-11eb-9bfc-a067f111949d.png)
![image](https://user-images.githubusercontent.com/31742919/113518599-9ce04880-9587-11eb-8b17-eff88cae31b3.png)

Useful stuff:
[mm-save.zip](https://github.com/gonetz/GLideN64/files/6255147/mm-save.zip)
[hd-minimaps.zip](https://github.com/gonetz/GLideN64/files/6255149/hd-minimaps.zip)

**Comment by gonetz** [comment]
@GhostlyDark I see nothing wrong in you screen shots. Please indicate where the problem is.

**Comment by gonetz** [comment]
> The texture shift might be a bit unrelated to this issue, so if you prefer to have it tested later, there's no problem for me. Or else it can be combined with the current branch. Your call.

Sorry for the late reply. I fetched your per_pixel_shift_offset branch. I'll check the commits and test it.
Could you make a draft PR with that branch to get test build on appveyor?

**Comment by gonetz** [comment]
> Per pixel shift offset #2476

Thanks! I tested it a bit. Normal textures work correct, but HD textures for 3D objects does not work:
![GLideN64_ZELDA_MAJORA'S_MASK_011](https://user-images.githubusercontent.com/8983092/113572021-9f06dd80-9641-11eb-90c8-d3fbe8931728.png)

**Comment by GhostlyDark** [comment]
HD Minimap is shifted upwards. If you look at the very top, you'll see the shadow is missing. The very bottom has additional shadow added.

**Comment by standard-two-simplex** [comment]
> HD Minimap is shifted upwards. If you look at the very top, you'll see the shadow is missing. The very bottom has additional shadow added.

The problem is with the heuristic for texture rectangles. First, I was missing some high res ratio factors, so I added them. However, even with those the game doesn't render perfectly in HD.

Secondly, the game treats the texture in a weird way. It is set to bilinear filter, but `dsdx` is `1.0` and `dtdy` is `-1.0`, so there is a one to one correspondence between texels and screen pixels. As a result, the texture is rendered as with nearest neighbour but with the bilinear filter active. On top of that the vertical direction is inverted. 

The heuristic treats it as a bilinearly filtered texture, but this texture would benefit more from the nearest neighbour case. I don't think it is possible to handle all situations any game can produce correctly. The best I got are these two options:
- 01a83c1d Use the heuristic for nearest neighbour always. It renders the HD textures quite well but there are out of bounds fetch issues in the bilinear case (as in FZeroX or Pilotwings, see ). They can still be solved by the texture bounds option.
- 37a9f363 Use the updated heuristic taking HD ratios into account. This HD textures are slightly missplaced with respect to native but don't look as bad as before.

@GhostlyDark Please, check if you find them satisfactory.

**Comment by standard-two-simplex** [comment]
> Normal textures work correct, but HD textures for 3D objects do not work:

I think the problem was missing parentheses around WRAP(). Please, recheck with current changes.

**Comment by GhostlyDark** [comment]
01a83c1 isn't perfect, the minimap is still slightly misplaced but looks much better and the rest looks about fine. I'd prefer this.

37a9f36 has a number of issues. The beating heart is visibly offset from the others, the rupee icon on the bottom left has a weird shadow outline, the minimap is a bit too high up and to the right.

master
![image](https://user-images.githubusercontent.com/31742919/113786234-0784b600-9739-11eb-8bb1-e29f85f839a1.png)

01a83c1
![image](https://user-images.githubusercontent.com/31742919/113786259-11a6b480-9739-11eb-8451-9d9b7a24e035.png)

37a9f36
![image](https://user-images.githubusercontent.com/31742919/113786280-1c614980-9739-11eb-997a-6536528bdef5.png)

**Comment by weinerschnitzel** [comment]
Numbers on the HUD clock are getting clipped, too. Master looks good there

**Comment by gonetz** [comment]
> I think the problem was missing parentheses around WRAP(). Please, recheck with current changes.

Fixed, thanks!

**Comment by standard-two-simplex** [comment]
I put another build where master-like behaviour is reproduced. It barely moves the HD textures, but moves the original to the place where the HD is. Strangely, this is closer to the behaviour of native res.

To be honest, I don't know why the 0.5f is there or should be there. I introduced it to reproduce the behaviour of older code. I seriously doubt it is correct, but it seems to give the best picture.

**Comment by GhostlyDark** [comment]
It looks pretty good with that change. Only thing I noticed is a very thin line at one of the sides that isn't supposed to be there. In this instance, on the left.

![image](https://user-images.githubusercontent.com/31742919/113901666-5e3ccf00-97cf-11eb-8b7d-1d69c286be27.png)

**Comment by standard-two-simplex** [comment]
So what shall we do about this? I would go with 01a83c1, because it seems the most mathematically correct approach to me. It reproduces the same behaviour as the vertex shift in native.

It creates issues https://github.com/gonetz/GLideN64/issues/2097#issuecomment-566009182 and https://github.com/gonetz/GLideN64/issues/2097#issuecomment-564745894 in non-native, but they can be overcome by the texture coordinate bounds option.

**Comment by gonetz** [comment]
I'm terribly sorry for being silent last several days.
My work swamped me again.
Ok, let it be 01a83c1 as its problems "can be overcome by the texture coordinate bounds option."
I hope to find a time to test it this weekend.

**Comment by standard-two-simplex** [comment]
> I'm terribly sorry for being silent last several days.

Don't worry, take the time you need, there are more important things.

**Comment by gonetz** [comment]
@standard-two-simplex 
I updated my standard-two-simplex-use_screen_coordinates branch with commits up to 01a83c1.
I added hot-key for texture coordinate bounds option. It is very useful for testing this functionality.
I also created another branch use_screen_coordinates_master-like with your commit "Change heuristic to render texture rectangles : master-like behaviour". 

I was about to add standard-two-simplex-use_screen_coordinates branch to master, but then I found some issues in Pilot Wings.
"Texture coordinate bounds" option fixes issues with texture mapping but creates another issue with garbage in demo mode.
I recorded a video to explain the issues:
https://drive.google.com/file/d/1C5geFpLuz__6H9yr_6GAcZ5B6q1bMixp/view?usp=sharing
Can this garbage issue be fixed?

**Comment by gonetz** [comment]
Btw, Pilot Wings looks a bit better with  "Change heuristic to render texture rectangles : master-like behaviour" - it has less glitches with "texture coordinate bounds" option off:

![PW1](https://user-images.githubusercontent.com/8983092/114294690-7d3aaa00-9aca-11eb-9b60-093338888a9a.png)
![PW2](https://user-images.githubusercontent.com/8983092/114294693-80ce3100-9aca-11eb-8744-7c3c7b56d3a4.png)

**Comment by standard-two-simplex** [comment]
> Can this garbage issue be fixed?

It seems that the game uses four textured rectangle commands to draw black bars on the edges. It is stranged that they are textured, but it seems so.

These are the commands,
```
gDPTextureRectangle(0.000000, 319.000000, 222.000000, 240.000000, 1, 0.000000, 0.000000, 0.031250, 0.031250, 0)
gDPTextureRectangle(0.000000, 319.000000, 1.000000, 8.000000, 1, 0.000000, 0.000000, 0.031250, 0.031250, 0)
gDPTextureRectangle(0.000000, 10.000000, 8.000000, 222.000000, 1, 0.000000, 0.000000, 0.031250, 0.031250, 0)
gDPTextureRectangle(310.000000, 319.000000, 8.000000, 222.000000, 1, 0.000000, 0.000000, 0.031250, 0.031250, 0)
```

For the first call, the texture coordinates I would expect would range between `s: [0, 318*0.03125]  = [0, 9.9375]` and `t : [0, 17*0.03125]=[0,5625]` which I compute as texture bounds.

For some reason, with texture bounds off, the texture coordinates seem to be really big and the clamping creates the issue. I don't understand why the texture coordinates would be so big.

**Comment by gonetz** [comment]
> For some reason, with texture bounds off, the texture coordinates seem to be really big and the clamping creates the issue.

The garbage appears with exture bounds **on**

**Comment by gonetz** [comment]
I added hotkey to toggle 2D texrects in native resolution: 88fe3bef5c67

I also noticed tremendous difference in quality between 01a83c1d1be1 and 506e34167f (master-like behaviour)
with "2D texrects in native resolution" enabled:

![PilotWings-NativeResTexrect](https://user-images.githubusercontent.com/8983092/114302410-18477a00-9af3-11eb-9637-f7c357582c96.jpg)

**Comment by standard-two-simplex** [comment]
> The garbage appears with texture bounds on

Yes, I meant that withouth bounds it has big texture coordinates and with the bounds it is clamped to unintended boundaries. I do not understand why the texture coordinates should be big, but it seems to be the correct behaviour as the rectangles are black.

**Comment by gonetz** [comment]
@standard-two-simplex I merged #2470. Thanks for this feature!
Please close related issues, or give me the list of issue, which now can be closed.

**Comment by standard-two-simplex** [comment]
This are the issues related. https://github.com/gonetz/GLideN64/pull/2334#issuecomment-711362297
